src/Analyser/NodeScopeResolver.php

```diff
@@ -5641,15 +5641,15 @@ static function (): void {
 
 						if ($assignedTypeIsCompatible) {
 							$scope = $scope->assignExpression($var, $assignedExprType, $assignedNativeType);
-						} elseif ($scope->isDeclareStrictTypes()) {
```

Comment:
with the last commit I made PHPStan always do type-coersion, even if non-strict-types are used.

I think we can do that in the changed areas, because we are working on natively typed properties and these always convert whatever you put into them

Reply:
See the reason why we can't simply do this:

```php
class Foo
{

	public int $foo;

	public function doFoo(string $s): void
	{
		$this->foo = $s;
		assertType('int', $this->foo);
	}

	public function doBar(): void
	{
		$this->foo = 'foo';
		assertType('int', $this->foo);
	}

}
```

Reports:

```
15     Expected type int, actual: *NEVER*¨
21     Expected type int, actual: *NEVER*
```

On your PR.

I'm fine with the state of assigning from before all the `toCoercedArgumentType` changes when strict_types is 0. It simply looked like this: https://github.com/phpstan/phpstan-src/blob/e6dc705b29b90dcc5f9773377c05aecbfe9fba3a/src/Analyser/NodeScopeResolver.php#L5215

But `->toCoercedArgumentType(true)` should only be called and assigned with strict_types = 1.

Reply:
thanks for the pointer. just to double check: in PHPStan 2.0.12 we get this results for your test:

```
class Foo
{

	public int $foo;

	public function doFoo(string $s): void
	{
		$this->foo = $s;
		assertType('string', $this->foo); // should be int
	}

	public function doBar(): void
	{
		$this->foo = 'foo';
		assertType("'foo'", $this->foo); // should be int
	}

}
```

and this is good enough to get back to, even though its wrong for now, right?